cupy/core/fusion.py

```diff
@@ -340,240 +538,186 @@ def res(*args, **kwargs):
-                                    reduce_type=post_type,
-                                    preamble=submodule_code)
+            ufunc.name, in_dtypes, out_dtypes))
+
+    def call_elementwise(self, f, args, kwargs):
+        raise NotImplementedError(
+            'Fusion for elementwise-kernel is not implemented yet')
+
+    def call(self, f, args, kwargs):
+        if type(f) is core.ufunc:
+            return self.call_ufunc(f, args, kwargs)
+        if type(f) is core.ElementwiseKernel:
+            return self.call_elementwise(f, args, kwargs)
+        raise Exception('Can\'t convert from {} to FusionOp'.format(type(f)))
+
+    def get_submodules_code(self):
+        res = ''.join(self.preamble_set)
+        res += '\n'.join([_.code() for _ in self.submodules.values()])
+        return res
+
+    def emit_operation_code(self):
+        res = '// {} operations\n'.format(len(self.op_list))
+        res += ''.join(v.declaration() for v in self.local_list)
+        res += ''.join(op.declaration_args() for op in self.op_list)
+        res += ''.join(op.code() for op in self.op_list)
+        return res
+
+    def emit_premap_code(self, in_params, operation):
+        return_var = self.premap_ret
+        module_code = string.Template('''
+        __device__ ${return_ctype} _pre_map(${in_params}) {
+        ${operation};
+        return ${return_var};
+        }
+        ''').substitute(
+            return_ctype=_dtype_to_ctype[return_var.dtype],
+            in_params=', '.join('{} v{}'.format(_dtype_to_ctype[v.dtype],
+                                                v.index)
+                                for v in in_params),
+            operation=operation,
+            return_var=return_var)
+        return module_code
+
+    def emit_postmap_code(self, out_params, operation):
+        in_param = self.postmap_param
+        in_ctype = _dtype_to_ctype[in_param.dtype]
+        module_code = string.Template('''
+        __device__ void _post_map(${in_ctype} in, ${out_params}) {
+        ${in_param} = in;
+        ${operation};
+        }
+        ''').substitute(
+            in_ctype=in_ctype,
+            in_param='{} v{}'.format(in_ctype, in_param.index),
+            out_params=', '.join('{} &v{}'.format(_dtype_to_ctype[v.dtype],
+                                                  v.index)
+                                 for v in out_params),
+            operation=operation)
+        return module_code
+
+    def emit_fix_code(self, reduce_ctype, fixed_dtype, operation):
+        module_code = string.Template('''
+        __device__ ${fixed_ctype} _post_fix(${reduce_ctype} a) {
+        ${fixed_ctype} out0;
+        ${operation};
+        return out0;
+        }
+        ''').substitute(
+            reduce_ctype=reduce_ctype,
+            fixed_ctype=_dtype_to_ctype[fixed_dtype],
+            operation=operation)
+        return module_code
+
+    def get_fusion_from_types(self, func, in_dtypes, name):
```

Comment:
Is it meaningful to emphasize `from_types` in the method name? It seems not to me.